cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -926,29 +994,47 @@ private class BuiltinExpectCallInstruction extends CallInstruction {
   }
 }
 
+private class BuiltinExpectCallValueNumber extends ValueNumber {
+  BuiltinExpectCallInstruction instr;
+
+  BuiltinExpectCallValueNumber() { this.getAnInstruction() = instr }
+
+  ValueNumber getCondition() { result.getAnInstruction() = instr.getCondition() }
+
+  Operand getAUse() { result = instr.getAUse() }
+}
+
+private class LogicalNotValueNumber extends ValueNumber {
+  LogicalNotInstruction instr;
+
+  LogicalNotValueNumber() { this.getAnInstruction() = instr }
+
+  ValueNumber getUnary() { result.getAnInstruction() = instr.getUnary() }
+}
+
 /**
  * Holds if `left == right + k` is `areEqual` if `cmp` evaluates to `value`,
  * and `cmp` is an instruction that compares the value of
  * `__builtin_expect(left == right + k, _)` to `0`.
  */
 private predicate builtin_expect_eq(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
+  CompareValueNumber cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
 ) {
-  exists(BuiltinExpectCallInstruction call, Instruction const, AbstractValue innerValue |
+  exists(BuiltinExpectCallValueNumber call, Instruction const, AbstractValue innerValue |
     int_value(const) = 0 and
     cmp.hasOperands(call.getAUse(), const.getAUse()) and
     compares_eq(call.getCondition(), left, right, k, areEqual, innerValue)
   |
-    cmp instanceof CompareNEInstruction and
+    cmp instanceof CompareNEValueNumber and
     value = innerValue
     or
-    cmp instanceof CompareEQInstruction and
+    cmp instanceof CompareEQValueNumber and
     value.getDualValue() = innerValue
   )
 }
 
 private predicate complex_eq(
-  CompareInstruction cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
+  ValueNumber cmp, Operand left, Operand right, int k, boolean areEqual, AbstractValue value
```

Comment:
Why not `CompareValueNumber`?

Reply:
We end up calling into the general `compares_eq` predicate, and that has cases where it's not a `CompareValueNumber`. For example in the case where it's a `LogicalNotValueNumber`. I'm actually surprised why this wasn't broken on `main` already, but I haven't been able to produce a counter-example 🤔 In any case, I understand why _this_ works so it seems slightly more safe